core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -165,13 +171,13 @@ public boolean add(T object) {
     // Locking this object protects against removing the exact object, but does not protect against
     // removing a distinct, but equivalent object.
     synchronized (object) {
-      if (!mObjects.addIfAbsent(object)) {
+      if (mPrimaryIndex.containsObject(object)) {
```

Comment:
Because both the check and add operations are protected by lock, I think the correctness will not be effected, because other thread will not be able to modify this object.